### PR TITLE
Dater implementation + rstests.

### LINKED ISSUE #67: Implement Dater
@DmitryKuzmenko mentioned he was starting this

### PR DESCRIPTION
**Pull request by DmitryKuzmenko**
In this commit I also used rstests. Looks very clear as for me.

Fixes #67 

Signed-off-by: Dmitry Kuzmenko <dmitry.kuzmenko@dsr-corporation.com>

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -18,9 +18,12 @@ lazy_static = "~1"
 sha2 = "~0.10"
 sha3 = "~0.10"
 thiserror = "~1"
+# Clippy fix: disable "oldtime" feature for chrono
```
Comment (qui-jason) [inline Cargo.toml:21]: Does this pass audit with the 'oldtime' removal?

Comment (qui-jason) [reply Cargo.toml:21]: I am hoping it does. When chrono is included with stock features it doesn't iirc.

File: src/core/dater.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    fn new_with_code_and_raw(code: &str, raw: &[u8]) -> Result<Matter> {
+        if !code.is_empty() {
+            validate_code(code)?;
+        }
+        if raw.is_empty() {
+            let qb64 = format!("{}{}", matter::Codex::DateTime.code(), now_as_b64());
+            Matter::new_with_qb64(&qb64)
+        } else {
+            Matter::new_with_code_and_raw(code, raw)
+        }
```
Comment (qui-jason) [inline src/core/dater.rs:59]: What happens here if code is empty and raw is not?

**Review by qui-jason — COMMENTED** [review]
Left a couple questions.

File: src/core/dater.rs (hunk 2 above)
Comment (qui-jason) [reply src/core/dater.rs:59]: Oh, you are relying on Matter to raise an error. I see.

File: src/core/dater.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+mod test_dater {
+    use super::{matter, Dater, Matter};
+    use rstest::rstest;
+
+    #[rstest]
+    fn test_new_default(
+        #[values(
+        &<Matter as Dater>::new_with_code_and_raw("", &[]).unwrap(),
+        &<Matter as Dater>::new_with_dts("").unwrap(),
+        &<Matter as Dater>::new_with_dtsb(b"").unwrap(),
```
Comment (qui-jason) [inline src/core/dater.rs:111]: Is this the equivalent of pytest's parametrize? https://docs.pytest.org/en/6.2.x/parametrize.html

If it is, amazing.

Comment (qui-jason) [reply src/core/dater.rs:111]: Yeah based on what I see it must be. rstest ftw.

File: Cargo.toml (hunk 1 above)
Comment (DmitryKuzmenko) [reply Cargo.toml:21]: Yes, that's correct
```
$ cargo audit && cargo clippy 
    Fetching advisory database from `https://github.com/RustSec/advisory-db.git`
      Loaded 509 security advisories (from /home/dimm/.cargo/advisory-db)
    Updating crates.io index
    Scanning Cargo.lock for vulnerabilities (220 crate dependencies)
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
```

File: src/core/dater.rs (hunk 2 above)
Comment (DmitryKuzmenko) [reply src/core/dater.rs:59]: Yep. Tests check this case.

**Review by qui-jason — COMMENTED** [review]
Okay I took a closer look and this looks okay to me but I will say with the caveat I haven't studied the python code. I'll do that now.

File: src/core/dater.rs (hunk 3 above)
Comment (DmitryKuzmenko) [reply src/core/dater.rs:111]: More or less. At least it's very easy and allows to build matrices multiplying `values` vectors. It's awesome.

**Comment by DmitryKuzmenko** [comment]
@qui-jason I've moved the logic from the python code. Just with one exception: I skipped `.datetime()` method for now. It's used in just 2 places in keripy and it's not more than `DateTime::parse_from_rfc3339(Dater.tds())`. Not sure we need it at least now.

File: Cargo.toml (hunk 1 above)
Comment (jasoncolburne) [reply Cargo.toml:21]: 🎉

**Review by m00sey — APPROVED** [review]

**Review by m00sey — COMMENTED** [review]
I didn't check for test vectors in keripy, I assume @jasoncolburne will, apologies.

**Review by jasoncolburne — APPROVED** [review]
Changes and test vectors look good to me. 100% diff coverage 🎉

**Pull request merged**